Improve the performance of indirect jump for T1C

qwe661234 opened the pull request:
Considering the T1C ends with the indirect jump, our interpreter mode now records the target of the indirect jump, identifying the most  frequent jump target for T1C. The T1 generated code for the indirect jump compares the selected jump target and executes the jump if the comparison results in equality. With this enhancement, T1C and T2C which are integrated later can proceed with an indirect jump if the target PC matches the selected jump target.

Based on the performance analysis, the benchmarks with significant number of indirect jump effectively improve the performance.

| Metric   | N_in_jmp  | Original | Propused | Speedup |
|----------|-----------|----------|----------|---------|
|miniz	   |    2098313|   1.266 s|   1.225 s|   +3.35%|
|sha512    |   10500727|   1.905 s|   1.861 s|   +2.36%|
|dhrystone |   20000618|   0.325 s|   0.253 s|  +28.46%|
|nqueens   |   44658722|   1.051 s|    0.79 s|  +33.04%|
|qsort	   |  275000250|   1.978 s|   1.517 s|  +30.39%|

jserv reviewed (changes requested):
Fix typo and improve the descriptions about the beneficial scenario in git commit messages.

jserv commented on src/rv32_template.c:
```diff
@@ -250,8 +275,9 @@ RVOP(
         alu32imm, 32, 0x81, 4, TMP, ~1U;
-        st, S32, TMP, PC;
         break;
+        bht;
```
`bht` might not be an intuitive name.

jserv commented on src/jit.c:
```diff
@@ -1336,6 +1336,25 @@ static void ra_load2_sext(struct jit_state *state,
+void parse_branch_history_table(struct jit_state *state, rv_insn_t *ir)
+{
+    int max_idx = 0;
+    for (int i = 0; i < HISTORY_SIZE; i++) {
```
Declare a variable which accesses `ir->branch_table`, so that you can write shorter code.

jserv commented on src/rv32_template.c:
```diff
@@ -107,6 +107,11 @@
  * |                                | store the result into dst.             |
  * | cond, src;                     | set condition if (src)                 |
  * | end;                           | set the end of condition if (src)      |
+ * | parsebt;                       | parse the branch table of indirect     |
```
Avoid using "parse", which might confuse. The intention was related to branch prediction. How about something like `predict`?

qwe661234 replied:
Ok, I think predict is better.

qwe661234 committed:
Improve the performance of indirect jump for T1C

Considering the T1C ends with the indirect jump, our interpreter mode
now records the target of the indirect jump, identifying the most
frequent jump target for T1C. The T1 generated code for the indirect
jump compares the selected jump target and executes the jump if the
comparison results in equality. With this enhancement, T1C and T2C
which are integrated later can proceed with an indirect jump if the
target PC matches the selected jump target.

Based on the performance analysis, the benchmarks with significant
number of indirect jump effectively improve the performance.

| Metric   | N_in_jmp  | Original | Propused | Speedup |
|----------+-----------+----------+----------+---------|
|miniz	   |    2098313|   1.266 s|   1.225 s|   +3.35%|
|sha512    |   10500727|   1.905 s|   1.861 s|   +2.36%|
|dhrystone |   20000618|   0.325 s|   0.253 s|  +28.46%|
|nqueens   |   44658722|   1.051 s|    0.79 s|  +33.04%|
|qsort	   |  275000250|   1.978 s|   1.517 s|  +30.39%|
